### DISCUSSION TITLE
Effects of, and optimal settings for, mismatch ratios

### DISCUSSION THREAD

**Discussion by hyanwong** [tskit-dev/tsinfer#563] [General]
In the [tsdate preprint](https://www.biorxiv.org/content/10.1101/2021.02.16.431497v2), we show a few plots of what happens when you adjust the mismatch ratio values when running the new `tsinfer` version. We conclude that for human-like datasets, mismatch ratios of 1 in both match samples and match ancestors passes gives a reasonable result. However, this may not always be the optimal value: it will be dependent on the underlying error model, and may differ between organisms. This discussion is aimed at collecting extra information about these mismatch parameters. FYI, here's the most relevant plot in the current preprint, based on 1500 samples and 1000-genomes-project-like error:

![mismatch_parameter_chr20_simulated_err-2](https://user-images.githubusercontent.com/4699014/126794276-2e01f7c6-a3eb-4c96-aade-7d5a02e142c7.jpg)

Which is created by running [test_mismatch_rates.py](https://github.com/hyanwong/tsinfer-benchmarking/blob/main/analysis/test_mismatch_rates.py)

**Comment by hyanwong** [comment]
There are a few extra metrics that we could plot:

1. "Generalised" RF distances, such as Martin Smith's MIC (Mutual Clustering Information metric).
2. The accuracy of dating mutations when the inferred TS is then dated using tsdate
3. The edge+mutation count when mutations are re-layed on the trees using parsimony (either with or without specifying the ancestral state)
4. The proportion of sites at which the focal ancestor is actually present in the tree

I've tried all of these on the standard OOA simulated dataset above with the following results:

1. MIC distance - seems to be basically the same as the RF metric but takes longer to calculate. Here's a example with only 150 samples, same 1000G error
<img width="368" alt="Screenshot 2021-07-23 at 19 00 34" src="https://user-images.githubusercontent.com/4699014/126822728-b355e0c2-346e-4b4f-9ef9-8710166e25fb.png">
2. Accuracy of further tsdate inference. This is pretty indirect, but very applicable to our case. In the 1500 sample case with 1000G error, the plots suggest that we should have a higher match samples ratio and a lower match ancestors ratio (here a *higher* correlation coefficient is better, so the optimal values are mismatch ratio = 1e-3 for match ancestors and 1e3 for match samples, which does make some sense.
<img width="559" alt="Screenshot 2021-07-23 at 19 46 23" src="https://user-images.githubusercontent.com/4699014/126827634-298325c6-d392-494a-9705-3bc10ece54ca.png">
3. Relaying using parsimony - if we do this, the number of mutations required when there is a high match ancestors mismatch ratio, seems relatively static, so that we can crank up the ancestors mismatch ratio which drops the number of edges but has little effect on the number of mutations so that by the measure of edges+mutations, a high match-ancestors mismatch ratio is beneficial (which seems odd)
<img width="275" alt="Screenshot 2021-07-23 at 19 51 29" src="https://user-images.githubusercontent.com/4699014/126828160-976870cf-f2a8-4c3e-9b5f-acc8585e8aaf.png">
4. Surprisingly often, at a focal site, we don't actually include the ancestor on which we expected the focal mutation to occur. We can measure this by looking at the proportion of sites that have the "focal" ancestor in the tree at the site position. Here's how that looks on the parameter map (lower mismatch ratios give more focal ancestors, as might be expected, but there's a jump around mismatch ratios of 10e-3)
<img width="643" alt="Screenshot 2021-07-24 at 15 34 57" src="https://user-images.githubusercontent.com/4699014/126871788-53ec34f0-0b0e-4042-92ff-71fa1e6ef0b0.png">

**Reply by jeromekelleher** [reply]
(Spotted a typo: need a p in "prootion" for fig title)

**Reply by hyanwong** [reply]
Thanks @jeromekelleher - corrected now. I don't think we want to add this particular plot to the tsdate paper though, do we? It's just to show in this discussion for further discussion

**Comment by hyanwong** [comment]
I now think that measuring the quality of imputation is the best way to go for this. See https://github.com/tskit-dev/tsinfer/issues/652

**Comment by moshejasper** [comment]
I'm working on tree sequence inference in a not-that-model organism with quite messy data (statistically phased Ae. aegypti short-reads). Lots of repetitive elements. Lots of missingness, not-that-great phasing, etc. 

For a lot of these reasons, I'm very keen to use recombination rates & mismatch to try and improve the original (somewhat suboptimal) trees. 

But I've read in random github issues that allowing these parameters on one of the two directions (either inferring ancestors or matching descendants afterwards) is very inadvisable. Indeed, my early runs a year ago suggested that under most normal settings, adding in mismatch created a situation where I got basically dozens of repeat mutations everywhere instead of a decent attempt to resolve things into reasonably parsimonious recombinations

Could you point me to a clear treatment of this, and advise whether the default parameters allow multiple mutations in both directions, & if so, which direction should I be being more careful about? 

Thanks!

Moshe

**Reply by hyanwong** [reply]
I'm afraid we have no clear recommendations / treatment yet. The HMM results of mismatch are pretty well-defined, but the implication in terms of quality of inference much less so. In `tsinfer` v 0.4 we changes the default in the .infer() method to only use mismatch during the match_samples phase, which is currently what we recommend. The effects of using it in math match_ancestors phase are not so clear.

The default setting of a mismatch ratio of 1 is almost certainly too high. I would probably set it to ~0.01 to start with (just a gut feeling though). If you have a (presumably simulated) ground truth, to which you can add realistic errors, we would be very interested to hear whether tuning the mismatch ratio parameter allows you to detect known problems.

It may be worth running an iterative tsinfer->tsdate->tsinfer pathway, where the date estimates are used to order haplotypes and re-infer.

A project to investigate the "best" mismatch values to use in different scenarios, perhaps using imputation to check on accuracy, would be very useful. I started some tests in this vein, but never got anywhere, sorry.

**Reply by moshejasper** [reply]
Thanks very much for this input! 
I'm going to try some runs with 0.01 and 0.0001 (& my default recombination ratio) & see how they turn out. 

My project is very much interested in this problem space, though I am perhaps still a little ways away from the really refined tests I feel will be needed. 

My current strategy is to vary inference parameters & data filtering stringency, etc. then look to identify where it appears to have made a substantial difference to the inferred tree sequence - but then also 'ground truth' in a rough way via comparing tree-sequenced derived metrics to equivalent 'traditional' metrics that loosely line up to the same thing. 

I think in the long run it would be great to add in imputation... presumably the idea would be to take high-quality data & then apply missingness in a profile that roughly matches my 'typical' missingness profile then look for accuracy of the reconstruction? 

(I'm also examining effects of pop structure & default allele dating, but increasingly getting the sense I may have bitten off more than I can chew!)

**Reply by hyanwong** [reply]
> presumably the idea would be to take high-quality data & then apply missingness in a profile that roughly matches my 'typical' missingness profile then look for accuracy of the reconstruction?

Exactly.

My gut feeling from talking to other people is that one of the main sources of error is likely to be wrongly polarised ancestral states. It might be worth focussing some efforts on investigating this in your dataset. 

By the way, you might be interested in https://github.com/tskit-dev/tsinfer/issues/670 for "sparsely triallelic" sites, which I suspect might be common in your dataset.

**Reply by moshejasper** [reply]
Hi. Thanks for the ref to #670. I found it a while ago and it has really been helpful in preserving a really important & informative subset of my data. Still working on extending that approach further if I can.

I'm very much interested in the polarization question, as it seems to be (imo) absolutely crucial to get it approximately right for things to function (I'm currently contrasting 'naive' methods with est-sfs & another approach to see if I get much difference. My dataset in many ways exacerbates the issue as there is substantial population structure present as well, meaning a frequency-based allelic age isn't going to be (always) informative (and I'm not fully on top of how this is going to affect these other programs either). I guess the hope is that by the end of this I can find a somewhat sane pathway through the treeseq weeds for people in a more 'non-model' organism space. Always keen for more convos & ideas about how to sharpen this along the way!
